torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -333,11 +333,34 @@ class ProcessGroupNCCL : public ProcessGroup {
       return !value_.isNone();
     }
 
+    std::shared_ptr<at::cuda::CUDAStream> getfutureNCCLCallbackStream() {
+      return futureNCCLCallbackStream_;
+    }
+
+    bool hasRecordStreamCallback() const {
+        return record_stream_cb_ != nullptr;
+    }
+
+    void setRecordStreamCallback(
+        std::function<void(const at::IValue&, at::cuda::CUDAStream)>
+            record_stream_cb) {
+      record_stream_cb_ = std::move(record_stream_cb);
+    }
+
+    void recordStream() {
+      TORCH_CHECK(record_stream_cb_);
+      // Do not free the underlying data storage of value_ before its usage on
```

Comment:
cc @pritamdamania87 , is there anyway we could write a test that would fail without this record stream callback logic?

Reply:
I think it might be possible but tricky to do. It probably has to be a C++ test and interact with CUDACachingAllocator in a way that the caching allocator frees up the data, but we're still using the data in futureNCCLCallbackStream.

Reply:
Currently I hacked the test file to locally verify that in a test case like fut.then(mult).then(div), CUDACachingAllocator::recordStream has been invoked by the first two futures, respectively.

Agree that there could be a better approach to create a C++ unit test that can allow CUDACachingAllocator to free up the cached data.